### PR TITLE
Improve genesis service

### LINKED ISSUE #1178: Don't count invalid deposits
## Description

Apparently, lighthouse counts deposits regardless of their validity. Don't count them unless they are valid.

## Version

`master` @ ddd63c0de146133ce7877bee218710a2b41defd8

## Present Behaviour

```
❯ lighthouse bn --http \
--testnet-dir ~/.opt/goerli/witti/lighthouse \
--eth1-endpoint http://127.0.0.1:8545

May 19 14:53:27.507 WARN Ethereum 2.0 is pre-release. This software is experimental.
May 19 14:53:27.508 INFO Data directory initialised              datadir: /home/user/.lighthouse
May 19 14:53:27.536 INFO Waiting for eth2 genesis from eth1      deposit_contract: 0x6225f431644ecf8a30b672d42b77c28297542b13, contract_deploy_block: 2723358, eth1_endpoint: http://127.0.0.1:8545, service: beacon
May 19 17:13:32.804 INFO Imported deposit log(s)                 new: 32, total: 32, latest_block: 2723546, service: beacon
May 19 17:13:46.837 INFO Imported deposit log(s)                 new: 32, total: 32, latest_block: 2723546, service: beacon
May 19 17:58:00.246 INFO Imported deposit log(s)                 new: 1, total: 33, latest_block: 2723724, service: beacon
May 19 17:58:14.290 INFO Imported deposit log(s)                 new: 32, total: 64, latest_block: 2723725, service: beacon
May 19 17:58:28.320 INFO Imported deposit log(s)                 new: 31, total: 64, latest_block: 2723725, service: beacon
May 19 18:19:17.820 INFO Imported deposit log(s)                 new: 2, total: 66, latest_block: 2723809, service: beacon
May 19 18:19:31.875 INFO Imported deposit log(s)                 new: 3, total: 67, latest_block: 2723810, service: beacon
May 19 18:19:45.906 INFO Imported deposit log(s)                 new: 15, total: 81, latest_block: 2723811, service: beacon
May 19 18:19:59.937 INFO Imported deposit log(s)                 new: 15, total: 82, latest_block: 2723812, service: beacon
May 19 18:20:13.966 INFO Imported deposit log(s)                 new: 15, total: 96, latest_block: 2723813, service: beacon
May 19 18:20:13.966 INFO Minimum genesis deposit count met       block_number: 2723813, deposit_count: 96, service: beacon
May 19 18:20:28.007 INFO Imported deposit log(s)                 new: 14, total: 96, latest_block: 2723813, service: beacon
May 19 23:11:23.400 INFO Imported deposit log(s)                 new: 1, total: 97, latest_block: 2724977, service: beacon
May 19 23:11:37.471 INFO Imported deposit log(s)                 new: 1, total: 97, latest_block: 2724977, service: beacon
May 20 00:50:58.651 INFO Imported deposit log(s)                 new: 1, total: 98, latest_block: 2725375, service: beacon
May 20 00:51:12.699 INFO Imported deposit log(s)                 new: 1, total: 98, latest_block: 2725375, service: beacon
May 20 00:55:11.757 INFO Imported deposit log(s)                 new: 1, total: 99, latest_block: 2725392, service: beacon
May 20 00:55:25.802 INFO Imported deposit log(s)                 new: 1, total: 99, latest_block: 2725392, service: beacon
```

The line `Minimum genesis deposit count met` indicates that the client might believe those deposits are all valid. However, it turns out that at least 32 are invalid: https://gist.github.com/ajsutton/8e03db5aea6e094d16333a7a0a0e5e39

I don't think you accept invalid deposits, it's probably just the counting happening before the validation.

## Expected Behaviour

Only count valid deposits.

## Steps to resolve

Yes.

### LINKED ISSUE #1186: Witti: Did not trigger genesis
```
~/.opt/sigp/lighthouse master* 1h 19m 18s
❯ lighthouse bn --http \
--testnet-dir ~/.opt/goerli/witti/lighthouse \
--eth1-endpoint http://127.0.0.1:9779
May 22 21:09:51.214 WARN Ethereum 2.0 is pre-release. This software is experimental.
May 22 21:09:51.214 INFO Data directory initialised              datadir: /home/user/.lighthouse
May 22 21:09:51.228 INFO Waiting for eth2 genesis from eth1      deposit_contract: 0x9eed6a5741e3d071d70817bed551d0078e9a2706, contract_deploy_block: 2740461, eth1_endpoint: http://127.0.0.1:9779, service: beacon
May 22 21:09:58.317 INFO Imported deposit log(s)                 new: 8, total: 8, latest_block: 2741153, service: beacon            
May 22 21:09:58.317 INFO Minimum genesis deposit count met       block_number: 2740838, deposit_count: 6, service: beacon            
May 22 21:09:58.376 INFO Inspected new eth1 block                block_number: 2740838, msg: did not trigger genesis, service: beacon
May 22 21:09:58.376 INFO Inspected new eth1 block                block_number: 2740839, msg: did not trigger genesis, service: beacon
May 22 21:09:58.376 INFO Inspected new eth1 block                block_number: 2740840, msg: did not trigger genesis, service: beacon
May 22 21:09:58.376 INFO Inspected new eth1 block                block_number: 2740841, msg: did not trigger genesis, service: beacon
May 22 21:09:58.377 INFO Inspected new eth1 block                block_number: 2740842, msg: did not trigger genesis, service: beacon
May 22 21:10:05.451 INFO Inspected new eth1 block                block_number: 2740843, msg: did not trigger genesis, service: beacon
May 22 21:10:05.451 INFO Inspected new eth1 block                block_number: 2740844, msg: did not trigger genesis, service: beacon
May 22 21:10:05.452 INFO Inspected new eth1 block                block_number: 2740845, msg: did not trigger genesis, service: beacon
May 22 21:10:05.452 INFO Inspected new eth1 block                block_number: 2740846, msg: did not trigger genesis, service: beacon
May 22 21:10:05.452 INFO Inspected new eth1 block                block_number: 2740847, msg: did not trigger genesis, service: beacon
```

Genesis was at 21:00:00 CEST, Prysm and Teku already chewing blocks.

master @ ea56dcb179d66b21f5e0af3bd6a3f62fc04d6736

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #1186
- Resolves #1178

## Proposed Changes

Applies the following to `Eth1GenesisService`:

- Fixes a bug where genesis was being triggered too late.
  - The candidate `BeaconState` was using `eth1_block.timestamp` as the genesis time, not `eth2_genesis_time(eth1_block.timestamp)`.
- Increases polling speed when we're trying to catch up to the eth1 head
  - Previously all polling happened at 7s intervals, this made it quite slow to catch up to the head.
- Improves logging UX.
- Adds an `Eth1GenesisService::statistics` method which returns statistics useful for future HTTP API upgrades.
- Removes unnecessary clones.

## TODO

- [x] Obtain genesis state roots from other Witti clients and ensure our root matches. (We match Prysm and Teku)

### CODE REVIEW

**Comment by paulhauner** [comment]
Note to self: this isn't working nicely at the moment because once we get close to the head of the chain we're polling super quick. We should update the slowdown logic so that it checks the follow distance against the current time, not genesis time (I think). This is a job for tomorrow.

File: beacon_node/genesis/src/eth1_genesis_service.rs [hunk 1]
```diff
@@ -106,67 +125,122 @@ impl Eth1GenesisService {
...
                         log,
-                        "Minimum genesis deposit count met";
-                        "deposit_count" => min_genesis_active_validator_count,
-                        "block_number" => viable_eth1_block,
+                        "Importing eth1 blocks";
                     );
-                    self.core.set_lowest_cached_block(viable_eth1_block);
-                    *sync_blocks = true
+                    self.eth1_service.set_lowest_cached_block(viable_eth1_block);
+                    sync_blocks = true
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/genesis/src/eth1_genesis_service.rs:141]: It's no longer a `RwLock`, if that makes a difference?

File: beacon_node/genesis/src/eth1_genesis_service.rs [hunk 2]
```diff
@@ -300,84 +365,76 @@ impl Eth1GenesisService {
...
+            Eth1Data {
+                block_hash: Hash256::zero(),
+                deposit_root: Hash256::zero(),
+                deposit_count: 0,
+            },
+            &spec,
+        );
+
+        state.genesis_time = eth2_genesis_time(eth1_block.timestamp, spec)
+            .map_err(|e| format!("Unable to set genesis time: {:?}", e))?;
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/genesis/src/eth1_genesis_service.rs:391]: Yep that's tidier, thanks

**Comment by paulhauner** [comment]
> Just minor changes requested. I'd also like to hear word that this produces the correct Witti genesis state

Thanks @michaelsproul, all comments addressed!

File: beacon_node/genesis/src/eth1_genesis_service.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/genesis/src/eth1_genesis_service.rs:141]: Ah nah I just meant the lack of semicolon, but it doesn't matter

**Comment by paulhauner** [comment]
I just made some adjustments, sorry:

- Remove estimated wait time when waiting for eth1 timestamp. It was too fiddly to prioritise at the moment.
- Added logging whilst waiting for genesis (after being concerned we had a deadlock after not seeing any logs).

**Pull request merged**